src/main/java/jenkins/branch/OrganizationFolder.java

```diff
@@ -0,0 +1,247 @@
+                    return listener;
+                }
+                @Override public SCMSourceObserver.ProjectObserver observe(final String projectName) {
+                    if (projects.containsKey(projectName)) {
+                        throw new IllegalArgumentException("Will not create duplicate subproject named " + projectName);
+                    }
+                    return new ProjectObserver() {
+                        List<SCMSource> sources = new ArrayList<SCMSource>();
+                        @Override public void addSource(SCMSource source) {
+                            sources.add(source);
+                            source.setOwner(OrganizationFolder.this);
+                        }
+                        @Override public void addAttribute(String key, Object value) throws IllegalArgumentException, ClassCastException {
+                            throw new IllegalArgumentException();
+                        }
+                        @Override public void complete() throws IllegalStateException, InterruptedException {
+                            if (sources == null) {
+                                throw new IllegalStateException();
+                            }
+                            for (MultiBranchProjectFactory factory : projectFactories) {
+                                MultiBranchProject<?, ?> project;
+                                try {
+                                    project = factory.createProject(OrganizationFolder.this, projectName, sources, Collections.<String,Object>emptyMap(), listener);
+                                } catch (InterruptedException x) {
+                                    throw x;
+                                } catch (Exception x) {
+                                    x.printStackTrace(listener.error("Failed to create a subproject " + projectName));
+                                    continue;
+                                }
+                                if (project != null) {
+                                    List<BranchSource> branchSources = project.getSourcesList();
+                                    for (SCMSource source : sources) {
+                                        // TODO we probably need to define a BranchPropertyFactory
+                                        branchSources.add(new BranchSource(source, new DefaultBranchPropertyStrategy(new BranchProperty[0])));
+                                    }
+                                    projects.put(projectName, project);
+                                    break;
+                                }
+                            }
+                        }
+                    };
+                }
+                @Override public void addAttribute(String key, Object value) throws IllegalArgumentException, ClassCastException {
+                    throw new IllegalArgumentException();
+                }
+            });
+        }
+        return projects;
+    }
```

Comment:
I find it hard to get through the above code. All the anonymous inner classes etc. Would it be possible to compose it in a way that makes it a bit easier to see the moving parts?